File: src/SLAM/rcomponent/src/rcomponent.cpp
/*! \class RComponent
 *  \file RComponent.cpp
 *	\author Robotnik Automation S.L.L
 *	\version 0.1.0
 *	\date 2014
 *  \brief Class to define a standard and shared structure (attributes & methods) for all the components
 * 			Single thread by default
 *
 *   This program is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include <rcomponent/rcomponent.h>
#include <robotnik_msgs/State.h>

/*! \fn RComponent::RComponent()
 *  \brief Constructor by default
 *	\param hz as double, sets the desired frequency of the controlthread
 *	\param h as ros::NodeHandle, ROS node handle
*/

namespace rcomponent
{
RComponent::RComponent()
{
  constructed = false;
}

RComponent::RComponent(ros::NodeHandle h) : RComponent::RComponent(h, ros::NodeHandle("~"))
{
  // XXX: this constructor is left to not break legacy code
}

RComponent::RComponent(ros::NodeHandle h, std::string name) : RComponent::RComponent(h, ros::NodeHandle(h, name))
{
}

RComponent::RComponent(ros::NodeHandle h, ros::NodeHandle ph) : nh_(h), pnh_(ph)
{
  init(nh_, pnh_);
}

int RComponent::init(ros::NodeHandle h)
{
  return init(h, ros::NodeHandle("~"));
}

int RComponent::init(ros::NodeHandle h, std::string name)
{
  return init(h, ros::NodeHandle(h, name));
}

int RComponent::init(ros::NodeHandle h, ros::NodeHandle ph)
{
  // if (ros_initialized or initialized or running)
  // {
  //   RCOMPONENT_INFO("Already initialized");

  //   return INITIALIZED;
  // }
  // desired_freq_ = 0;

  nh_ = h;
  pnh_ = ph;

  // Realizar para cada una de las clases derivadas

  // name assignment
  component_name.assign(constructComponentNameFromHandles());

  // Set main flags to false
  ros_initialized = initialized = running = false;

  // reads params from server
  rosReadParams();

  if (desired_freq_ <= 0.0)
    desired_freq_ = DEFAULT_THREAD_DESIRED_HZ;

  state = robotnik_msgs::State::INIT_STATE;

  threadData.pthreadPar.prio = 25;               // Priority level 0[min]-80[max]
  threadData.pthreadPar.clock = CLOCK_REALTIME;  // 0-CLOCK_MONOTONIC 1-CLOCK_REALTIME
  constructed = true;

  t_state_transition_ = ros::Time::now();

  return OK;
}

/*! \fn RComponent::~RComponent()
 * Destructor by default
*/
RComponent::~RComponent()
{
  running = false;
  constructed = false;
}

/*! \fn int RComponent::setup()
 * Configures and initializes the component
 * \return OK
 * \return INITIALIZED if the component is already intialized
 * \return ERROR
*/
int RComponent::setup()
{
  // Checks if has been created with a fully qualified constructor, or with an empty
  // constructor but then init has been called
  if (constructed == false)
  {
    RCOMPONENT_ERROR("Not properly constructed");
    return ERROR;
  }

  // Checks if has been initialized
  if (initialized)
  {
    RCOMPONENT_INFO("Already initialized");

    return INITIALIZED;
  }

  //
  ///////////////////////////////////////////////////
  // Setups the component or another subcomponents if it's necessary //
  ///////////////////////////////////////////////////

  initialized = true;

  return OK;
}

/*! \fn int RComponent::shutDown()
 * Closes and frees the reserved resources
 * \return OK
 * \return ERROR if fails when closes the devices
 * \return RUNNING if the component is running
 * \return NOT_INITIALIZED if the component is not initialized
*/
int RComponent::shutdown()
{
  if (running)
  {
    RCOMPONENT_INFO("Impossible while thread running, first must be stopped");
    return THREAD_RUNNING;
  }
  if (!initialized)
  {
    RCOMPONENT_INFO("Impossible because of it's not initialized");
    return NOT_INITIALIZED;
  }

  //
  ///////////////////////////////////////////////////////
  // ShutDowns another subcomponents if it's necessary //
  ///////////////////////////////////////////////////////

  initialized = false;

  return OK;
}

/*! \fn int RComponent::start()
 * Starts the control operation of the component and its subcomponents in a blocking way
 * \return OK
 * \return RUNNING if it's already running
 * \return NOT_INITIALIZED if the component is not initialized
*/
int RComponent::start()
{
  if (running)
  {
    RCOMPONENT_INFO("Component's thread is already running");
    return THREAD_RUNNING;
  }

  // Performs ROS setup

  int setup_result = rosSetup();

  if (setup_result == rcomponent::ERROR)
    return rcomponent::ERROR;

  RCOMPONENT_INFO("Started");

  running = true;

  switchToState(robotnik_msgs::State::INIT_STATE);
  // Executes the control loop
  controlLoop();

  return OK;
}

/*! \fn int RComponent::asyncStart()
 * Starts the control thread of the component and its subcomponents
 * \return OK
 * \return RUNNING if it's already running
 * \return NOT_INITIALIZED if the component is not initialized
*/
int RComponent::asyncStart()
{
  if (running)
  {
    RCOMPONENT_INFO("Component's thread is already running");
    return THREAD_RUNNING;
  }

  // Performs ROS setup
  int setup_result = rosSetup();

  if (setup_result == rcomponent::ERROR)
    return rcomponent::ERROR;

  pthread_attr_t attr;  // Thread attributed for the component threads spawned in this function

  RCOMPONENT_DEBUG("Launching the thread");
  pthread_attr_init(&attr);
  pthread_attr_setdetachstate(&attr, PTHREAD_CREATE_DETACHED);

  running = true;

  switchToState(robotnik_msgs::State::INIT_STATE);
  if (pthread_create(&threadData.pthreadId, &attr, &RComponent::asyncControlLoop, this) != 0)
  {
    RCOMPONENT_ERROR("Could not create ControlThread");
    pthread_attr_destroy(&attr);
    running = false;
    return ERROR;
  }

  return OK;
}
/*! \fn int RComponent::stop()
 * Stops the control thread of the Motors
 * \return OK
 * \return ERROR if it can't be stopped
 * \return THREAD_NOT_RUNNING if the thread is not running
*/
int RComponent::stop()
{
  if (!running)
  {
    RCOMPONENT_INFO("Thread not running");

    return THREAD_NOT_RUNNING;
  }
  //
  ///////////////////////////////////////////////////
  // Stops another subcomponents, if it's necessary //
  ///////////////////////////////////////////////////
  //
  RCOMPONENT_INFO("Stopping the component");

  running = false;

  usleep(100000);

  return OK;
}

/*! \fn int RComponent::isRunning()
 *  Checks if component is running
 *  \return TRUE component is running
 *  \return FALSE component is not running
*/
bool RComponent::isRunning()
{
  return running;
}

void* RComponent::asyncControlLoop(void* context)
{
  ((RComponent*)context)->controlLoop();
  return NULL;
}

/*!	\fn void RComponent::controlLoop()
 *	\brief All core component functionality is contained in this thread.
*/
void RComponent::controlLoop()
{
  RCOMPONENT_INFO("Init");
  ros::Rate r(desired_freq_);
  ros::Time t1, t2;
  while (running && ros::ok())
  {
    t1 = ros::Time::now();

    switch (state)
    {
      case robotnik_msgs::State::INIT_STATE:
        initState();
        break;

      case robotnik_msgs::State::STANDBY_STATE:
        standbyState();
        break;

      case robotnik_msgs::State::READY_STATE:
        readyState();
        break;

      case robotnik_msgs::State::SHUTDOWN_STATE:
        shutdownState();
        break;

      case robotnik_msgs::State::EMERGENCY_STATE:
        emergencyState();
        break;

      case robotnik_msgs::State::FAILURE_STATE:
        failureState();
        break;
    }

    allState();

    ros::spinOnce();
    r.sleep();

    t2 = ros::Time::now();
    try
    {
      real_freq = 1.0 / (t2 - t1).toSec();
    }
    catch (std::runtime_error& ex)
    {
      RCOMPONENT_ERROR("Exception: [%s]", ex.what());
    }
  }

  shutdownState();
  // Performs ROS Shutdown
  rosShutdown();

  RCOMPONENT_INFO("End");
}

/*!	\fn void RComponent::initState()
 *	\brief Actions performed on initial
 * 	Setups the component
*/
void RComponent::initState()
{
  // If component setup is successful goes to STANDBY (or READY) state
  if (setup() != ERROR)
  {
    switchToState(robotnik_msgs::State::STANDBY_STATE);
  }
}

/*!	\fn void RComponent::shutdownState()
 *	\brief Actions performed on Shutdown state
*/
void RComponent::shutdownState()
{
  if (shutdown() == OK)
  {
    switchToState(robotnik_msgs::State::INIT_STATE);
  }
}

/*!	\fn void RComponent::standbyState()
 *	\brief Actions performed on Standby state
*/
void RComponent::standbyState()
{
}

/*!	\fn void RComponent::readyState()
 *	\brief Actions performed on ready state
*/
void RComponent::readyState()
{
}

/*!	\fn void RComponent::EmergencyState()
 *	\brief Actions performed on emergency state
*/
void RComponent::emergencyState()
{
}

/*!	\fn void RComponent::FailureState()
 *	\brief Actions performed on failure state
*/
void RComponent::failureState()
{
}

/*!	\fn void RComponent::AllState()
 *	\brief Actions performed on all states
*/
void RComponent::allState()
{
  rosPublish();
}

/*!	\fn double RComponent::getUpdateRate()
 * 	\brief Gets current update rate of the thread
 * 	\return real frequency of the thread
*/
double RComponent::getUpdateRate()
{
  return desired_freq_;
}

/*!	\fn int RComponent::getState()
 * 	\brief returns the state of the component
*/
int RComponent::getState()
{
  return state;
}

/*!	\fn char *RComponent::getStateString()
 *	\brief Gets the state of the component as char* string
*/
// char* RComponent::getStateString()
//{
//  return getStateString(state);
//}

/*!	\fn std::string RComponent::getStateString()
 *	\brief Gets the state of the component as std::string
*/
std::string RComponent::getStateString()
{
  return std::string(getStateString(state));
}

/*! \fn char *RComponent::getComponentName()
 *  \brief Returns the name of the component
*/
const char* RComponent::getComponentName()
{
  return component_name.c_str();
}

/*! \fn char *RComponent::setComponentName()
 *  \brief Returns the name of the component
*/
void RComponent::setComponentName(const std::string& name)
{
  component_name.assign(name);
}

/*! \fn std::string RComponent::constructComponentNameFromHandles()
 *  \brief Constructs a proper name for the component, using information from the handles
 *  passed to it
*/
std::string RComponent::constructComponentNameFromHandles()
{
  std::string name = "RComponent";
  if (pnh_.getNamespace() != "" and pnh_.getNamespace() != "~")
  {
    name = pnh_.getNamespace();
  }
  else if (nh_.getNamespace() != "" and nh_.getNamespace() != "~")
  {
    name = nh_.getNamespace();
  }

  size_t i = name.rfind('/', name.length());
  if (i != std::string::npos)
  {
    name = name.substr(i + 1, name.length() - i);
  }
  return name;
}

/*! \fn std::string RComponent::getPrivateNamespace()
 *  \brief Sets the public namespace of the component
*/
const std::string RComponent::getPrivateNamespace()
{
  return pnh_.getNamespace();
}

/*! \fn std::string RComponent::getPublicNamespace()
 *  \brief Returns the public namespace of the component
*/
const std::string RComponent::getPublicNamespace()
{
  return nh_.getNamespace();
}

/*!	\fn char *RComponent::getStateString(int state)
 *	\brief Gets the state as a string
*/
char* RComponent::getStateString(int state)
{
  switch (state)
  {
    case robotnik_msgs::State::INIT_STATE:
      return (char*)"INIT";
      break;
    case robotnik_msgs::State::STANDBY_STATE:
      return (char*)"STANDBY";
      break;
    case robotnik_msgs::State::READY_STATE:
      return (char*)"READY";
      break;
    case robotnik_msgs::State::EMERGENCY_STATE:
      return (char*)"EMERGENCY";
      break;
    case robotnik_msgs::State::FAILURE_STATE:
      return (char*)"FAILURE";
      break;
    case robotnik_msgs::State::SHUTDOWN_STATE:
      return (char*)"SHUTDOWN";
      break;
    default:
      return (char*)"UNKNOWN";
      break;
  }
}

/*!	\fn void RComponent::switchToState(int new_state)
 * 	function that switches the state of the component into the desired state
 * 	\param new_state as an integer, the new state of the component
*/
void RComponent::switchToState(int new_state)
{
  if (new_state == state)
    return;

  // saves the previous state
  previous_state = state;
  RCOMPONENT_INFO("%s -> %s", getStateString(state), getStateString(new_state));
  state = new_state;

  switch (state)
  {
    case robotnik_msgs::State::INIT_STATE:
      switchToInitState();
      break;
    case robotnik_msgs::State::STANDBY_STATE:
      switchToStandbyState();
      break;
    case robotnik_msgs::State::READY_STATE:
      switchToReadyState();
      break;
    case robotnik_msgs::State::EMERGENCY_STATE:
      switchToEmergencyState();
      break;
    case robotnik_msgs::State::FAILURE_STATE:
      switchToFailureState();
      break;
    case robotnik_msgs::State::SHUTDOWN_STATE:
      switchToShutdownState();
      break;
  }

  t_state_transition_ = ros::Time::now();
}

/*!	\fn void RComponent::switchToInitState()
 * 	callback executed when moving to init state
*/
void RComponent::switchToInitState()
{
}

/*!	\fn void RComponent::switchToStandbyState()
 * 	callback executed when moving to standby state
*/
void RComponent::switchToStandbyState()
{
}

/*!	\fn void RComponent::switchToReadyState()
 * 	callback executed when moving to ready state
*/
void RComponent::switchToReadyState()
{
}

/*!	\fn void RComponent::switchToEmergencyState()
 * 	callback executed when moving to emergency state
*/
void RComponent::switchToEmergencyState()
{
}

/*!	\fn void RComponent::switchToFailureState()
 * 	callback executed when moving to failure state
*/
void RComponent::switchToFailureState()
{
}

/*!	\fn void RComponent::switchToShutdownState()
 * 	callback executed when moving to shutdown state
*/
void RComponent::switchToShutdownState()
{
}

/*!	\fn void RComponent::rosSetup()
 * 	\brief Setups all ROS' stuff
*/
int RComponent::rosSetup()
{
  // Checks if has been initialized
  if (ros_initialized)
  {
    RCOMPONENT_INFO("Already initialized");

    return INITIALIZED;
  }

  state_publisher = pnh_.advertise<robotnik_msgs::State>("state", 1);
  // state_publisher = pnh_.advertise<std_msgs::Empty>("state", 1);

  ros_initialized = true;

  return OK;

  /*

  status_pub_ = pnh_.advertise<agvs_controller::DspicStatus>("status", 1);
  odom_pub_ = pnh_.advertise<nav_msgs::Odometry>(odom_frame_id_, 1);
  calibrate_srv_ = pnh_.advertiseService("calibrate",  &dspic_controller_node::CalibrateSrv, this);
  set_odom_service_ = pnh_.advertiseService("set_odometry", &dspic_controller_node::SetOdometry, this);*/
}

/*!	\fn void RComponent::rosReadParams
 * 	\brief Reads the params set in ros param server
*/
void RComponent::rosReadParams()
{
  pnh_.param("desired_freq", desired_freq_, DEFAULT_THREAD_DESIRED_HZ);

  /* Example
  pnh_.param<std::string>("port", port_, DEFAULT_DSPIC_PORT);
  pnh_.param<std::string>("odom_frame_id", odom_frame_id_, "/odom_diff");
  pnh_.param<std::string>("base_frame_id", base_frame_id_, "/base_link");
  pnh_.param("publish_tf", publish_tf_, false);
  pnh_.param("desired_freq", desired_freq_, desired_freq_);*/
}

/*!	\fn int RComponent::rosShutdown()
 * 	\brief Closes all ros stuff
*/
int RComponent::rosShutdown()
{
  if (running)
  {
    RCOMPONENT_INFO("Impossible while thread running, first must be stopped");
    return THREAD_RUNNING;
  }
  if (!ros_initialized)
  {
    RCOMPONENT_INFO("Impossible because of it's not initialized");
    return NOT_INITIALIZED;
  }

  ros_initialized = false;

  return OK;
}

/*!	\fn void RComponent::rosPublish()
 * 	\brief Reads data a publish several info into different topics
*/
void RComponent::rosPublish()
{
  robotnik_msgs::State msg;

  // STATE
  msg.state = this->state;
  msg.desired_freq = this->desired_freq_;
  msg.real_freq = this->real_freq;
  msg.state_description = getStateString();

  state_publisher.publish(msg);
}

/*!	\fn bool RComponent::checkTopicsHealth(std::string topic)
 * 	\brief Checks the topic health of all the subscribed topics or specific ones
 *  \param topic as std::string, topic to check. If empty all the topics are checked as a group
 *  \return true if health is ok, false otherwise
*/
bool RComponent::checkTopicsHealth(std::string topic_id)
{
  std::map<std::string, TopicHealthMonitor>::iterator it;

  if (data_health_monitors_.empty())
  {
    RCOMPONENT_WARN_STREAM_ONCE("Topics health monitor is empty");
    return true;
  }

  if (topic_id.empty() == false)
  {
    it = data_health_monitors_.find(topic_id);

    if (it != data_health_monitors_.end())
    {
      return it->second.isReceiving();
    }
    else
    {
      RCOMPONENT_WARN_STREAM_THROTTLE(5, "Topic " << it->first << " not being received");
      return false;
    }
  }
  else  // Only check required topics
  {
    for (it = data_health_monitors_.begin(); it != data_health_monitors_.end(); ++it)
    {
      if (it->second.isReceiving() == false && it->second.isRequired() == true)
      {
        RCOMPONENT_WARN_STREAM_THROTTLE(5, "Topic " << it->first << " not being received");
        return false;
      }
    }

    return true;
  }
}

/*!	\fn int RComponent::addTopicsHealth(ros::Subscriber *subscriber = 0, std::string topic_id, double timeout)
 * 	\brief Adds a topic health for the subscriber
 *  \param subscriber as ros::Subscriber*, pointer to the Subscriber to check health
 *  \param topic_id as std::string, topic id to associate with the subscriber. If empty it will use the full topic name
 *  \param timeout as std::string, topic id to associate with the subscriber. If empty it will use the full topic name
 *  \return 0 if ok, -1 otherwise
*/
int RComponent::addTopicsHealth(ros::Subscriber* subscriber, std::string topic_id, double timeout, bool required)
{
  std::string map_id;
  if (subscriber == NULL)
  {
    RCOMPONENT_ERROR("Subscriber to check is NULL");
    return -1;
  }

  if (topic_id.empty() == true)
  {
    map_id = subscriber->getTopic();
  }
  else
  {
    map_id = topic_id;
  }
  if (timeout <= 0)
  {
    RCOMPONENT_ERROR("timeout (%.lf) has to be >= 0", timeout);
    timeout = 1.0;
  }

  data_health_monitors_[map_id] = TopicHealthMonitor(subscriber, timeout, required);

  return 0;
}

/*!	\fn int RComponent::tickTopicsHealth(std::string topic_id)
 * 	\brief Ticks the selected topic to notify that is being received
 *  \param topic_id as std::string, topic id that identifies the subscriber
 *  \return 0 if ok, -1 otherwise
*/
int RComponent::tickTopicsHealth(std::string topic_id)
{
  if (topic_id.empty() == true)
  {
    RCOMPONENT_ERROR_STREAM_THROTTLE(5, "topic id " << topic_id << " cannot be empty");
    return -1;
  }

  std::map<std::string, TopicHealthMonitor>::iterator it;

  it = data_health_monitors_.find(topic_id);

  if (it != data_health_monitors_.end())
  {
    it->second.tick();
    return 0;
  }

  RCOMPONENT_ERROR_STREAM_THROTTLE(5, "topic id " << topic_id << " does not exist");
  return -1;
}

/*!	\fn ros::Duration RComponent::getElapsedTimeSinceLastStateTransition()
 * 	\brief Returns the elapsed time since the last state transition
*/
ros::Duration RComponent::getElapsedTimeSinceLastStateTransition()
{
  return ros::Time::now() - t_state_transition_;
}

/*!	\fn ros::Time RComponent::getElapsedTimeSinceLastStateTransition()
 * 	\brief Returns the state transition time
*/
ros::Time RComponent::getStateTransitionTime()
{
  return t_state_transition_;
}

}  // namespace rcomponent

File: src/SLAM/summit/rcomponent/src/rcomponent/rcomponent.py
#!/usr/bin/env python

# Software License Agreement (BSD License)
#
# Copyright (c) 2015, Robotnik Automation SLL
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
#  * Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
#  * Redistributions in binary form must reproduce the above
#    copyright notice, this list of conditions and the following
#    disclaimer in the documentation and/or other materials provided
#    with the distribution.
#  * Neither the name of Robotnik Automation SSL nor the names of its
#    contributors may be used to endorse or promote products derived
#    from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

import rospy

import time
import threading

try:
    from topic_health_monitor import TopicHealthMonitor
except ImportError:
    from rcomponent.topic_health_monitor import TopicHealthMonitor

from robotnik_msgs.msg import State

DEFAULT_FREQ = 10.0

# Class Template of Robotnik component for Pyhton


class RComponent:

    def __init__(self):
        self._node_name = rospy.get_name()
        self.ros_read_params()

        self._real_freq = 0.0
        # Saves the state of the component
        self._state = State.INIT_STATE
        # Saves the previous state
        self._previous_state = State.INIT_STATE
        # flag to control the initialization of the component
        self._initialized = False
        # flag to control the initialization of ROS stuff
        self._ros_initialized = False
        # flag to control that the control loop is running
        self._running = False
        # Variable used to control the loop frequency
        self._time_sleep = 1.0 / self._desired_freq
        # State msg to publish
        self._msg_state = State()
        # Timer to publish state
        self._publish_state_timer = 1

        self._t_publish_state = threading.Timer(self._publish_state_timer, self.publish_ros_state)
        # to save the time of the last state transition
        self._t_state_transition = rospy.Time(0)
        # dict to save all the topic health monitor objects
        self._data_health_monitors = {}

    def ros_read_params(self):
        '''
            Gets params from param server
        '''

        try:
            self._desired_freq = rospy.get_param('~desired_freq', default=DEFAULT_FREQ)
        except rospy.ROSException as e:
            rospy.logerr('%s' % (e))
            exit(-1)

        # Checks value of freq
        if self._desired_freq <= 0.0:
            rospy.loginfo('%s::init: Desired freq to %f is not possible. Setting _desired_freq to %f' %
                          (self._node_name, self._desired_freq, DEFAULT_FREQ))
            self._desired_freq = DEFAULT_FREQ

    def setup(self):
        '''
                Initializes de hand
                @return: True if OK, False otherwise
        '''
        self._initialized = True

        return 0

    def ros_setup(self):
        '''
                Creates and inits ROS components
        '''
        if self._ros_initialized:
            return 0

        # Publishers
        self._state_publisher = rospy.Publisher('~state', State, queue_size=10)
        # Subscribers
        # topic_name, msg type, callback, queue_size
        # self.topic_sub = rospy.Subscriber('topic_name', Int32, self.topic_cb, queue_size = 10)
        # Service Servers
        # self.service_server = rospy.Service('~service', Empty, self.service_cb)
        # Service Clients
        # self.service_client = rospy.ServiceProxy('service_name', ServiceMsg)
        # ret = self.service_client.call(ServiceMsg)

        self._ros_initialized = True

        self.publish_ros_state()

        return 0

    def shutdown(self):
        '''
                Shutdowns device
                @return: 0 if it's performed successfully, -1 if there's any problem or the component is running
        '''
        if self._running or not self._initialized:
            return -1
        rospy.loginfo('%s::shutdown' % self._node_name)

        # Cancels current timers
        self._t_publish_state.cancel()

        self._state_publisher.unregister()

        self._initialized = False

        return 0

    def ros_shutdown(self):
        '''
                Shutdows all ROS components
                @return: 0 if it's performed successfully, -1 if there's any problem or the component is running
        '''
        if self._running or not self._ros_initialized:
            return -1

        # Performs ROS topics & services shutdown
        self._state_publisher.unregister()

        self._ros_initialized = False

        return 0

    def stop(self):
        '''
                Creates and inits ROS components
        '''
        self._running = False

        return 0

    def start(self):
        '''
                Runs ROS configuration and the main control loop
                @return: 0 if OK
        '''
        self.ros_setup()

        if self._running:
            return 0

        self._running = True

        self.control_loop()

        return 0

    def control_loop(self):
        '''
                Main loop of the component
                Manages actions by state
        '''

        while self._running and not rospy.is_shutdown():
            t1 = time.time()

            if self._state == State.INIT_STATE:
                self.init_state()

            elif self._state == State.STANDBY_STATE:
                self.standby_state()

            elif self._state == State.READY_STATE:
                self.ready_state()

            elif self._state == State.EMERGENCY_STATE:
                self.emergency_state()

            elif self._state == State.FAILURE_STATE:
                self.failure_state()

            elif self._state == State.SHUTDOWN_STATE:
                self.shutdown_state()

            self.all_state()

            t2 = time.time()
            tdiff = (t2 - t1)

            t_sleep = self._time_sleep - tdiff

            if t_sleep > 0.0:
                try:
                    rospy.sleep(t_sleep)
                except rospy.exceptions.ROSInterruptException:
                    rospy.loginfo('%s::control_loop: ROS interrupt exception' % self._node_name)
                    self._running = False

            t3 = time.time()
            self._real_freq = 1.0/(t3 - t1)

        self._running = False
        # Performs component shutdown
        self.shutdown_state()
        # Performs ROS shutdown
        self.ros_shutdown()
        rospy.loginfo('%s::control_loop: exit control loop' % self._node_name)

        return 0

    def ros_publish(self):
        '''
                Publish topics at standard frequency
        '''

        return 0

    def init_state(self):
        '''
                Actions performed in init state
        '''

        if not self._initialized:
            self.setup()

        else:
            self.switch_to_state(State.STANDBY_STATE)

        return

    def standby_state(self):
        '''
                Actions performed in standby state
        '''
        self.switch_to_state(State.READY_STATE)

        return

    def ready_state(self):
        '''
                Actions performed in ready state
        '''

        return

    def shutdown_state(self):
        '''
                Actions performed in shutdown state
        '''
        if self.shutdown() == 0:
            self.switch_to_state(State.INIT_STATE)

        return

    def emergency_state(self):
        '''
                Actions performed in emergency state
        '''

        return

    def failure_state(self):
        '''
                Actions performed in failure state
        '''

        return

    def switch_to_state(self, new_state):
        '''
                Performs the change of state
        '''

        if self._state == new_state:
            return

        if new_state == State.INIT_STATE:
            self.switch_to_init_state()
        elif new_state == State.STANDBY_STATE:
            self.switch_to_standby_state()
        elif new_state == State.READY_STATE:
            self.switch_to_ready_state()
        elif new_state == State.EMERGENCY_STATE:
            self.switch_to_emergency_state()
        elif new_state == State.FAILURE_STATE:
            self.switch_to_failure_state()
        elif new_state == State.SHUTDOWN_STATE:
            self.switch_to_shutdown_state()

        self._previous_state = self._state
        self._state = new_state
        rospy.loginfo('%s::switch_to_state: from %s to %s' %
                      (self._node_name, self.state_to_string(self._previous_state), self.state_to_string(self._state)))

        self._t_state_transition = rospy.Time.now()

        return

    def switch_to_init_state(self):
        '''
            Function called during the transition to init_state
        '''
        pass

    def switch_to_standby_state(self):
        '''
            Function called during the transition to standby_state
        '''
        pass

    def switch_to_ready_state(self):
        '''
            Function called during the transition to ready_state
        '''
        pass

    def switch_to_emergency_state(self):
        '''
            Function called during the transition to emergency_state
        '''
        pass

    def switch_to_failure_state(self):
        '''
            Function called during the transition to failure_state
        '''
        pass

    def switch_to_shutdown_state(self):
        '''
            Function called during the transition to shutdown_state
        '''
        pass

    def all_state(self):
        '''
                Actions performed in all states
        '''
        self.ros_publish()

        return

    def state_to_string(self, state):
        '''
                @param state: state to set
                @type state: State
                @returns the equivalent string of the state
        '''
        if state == State.INIT_STATE:
            return 'INIT_STATE'

        elif state == State.STANDBY_STATE:
            return 'STANDBY_STATE'

        elif state == State.READY_STATE:
            return 'READY_STATE'

        elif state == State.EMERGENCY_STATE:
            return 'EMERGENCY_STATE'

        elif state == State.FAILURE_STATE:
            return 'FAILURE_STATE'

        elif state == State.SHUTDOWN_STATE:
            return 'SHUTDOWN_STATE'
        else:
            return 'UNKNOWN_STATE'

    def publish_ros_state(self):
        '''
                Publish the State of the component at the desired frequency
        '''
        self._msg_state.state = self._state
        self._msg_state.state_description = self.state_to_string(self._state)
        self._msg_state.desired_freq = self._desired_freq
        self._msg_state.real_freq = self._real_freq

        if rospy.is_shutdown() == False:
            self._state_publisher.publish(self._msg_state)
            self._t_publish_state = threading.Timer(self._publish_state_timer, self.publish_ros_state)
            self._t_publish_state.start()

    def get_state_transition_elapsed_time(self):
        '''
            @returns the elapsed time since the last state transition as rospy.Time.Duration
        '''
        return rospy.Time.now() - self._t_state_transition

    def get_state_transition_time(self):
        '''
            @returns the current value of the state transition time
        '''
        return self._t_state_transition

    def add_topics_health(self, subscriber, topic_id='', timeout=5.0, required=True):
        '''
            @brief Adds a topic health for the subscriber
            @param subscriber as a rospy.Subscriber to check
            @param topic_id as string. Id associated to the topic. If empty it will use the full topic name
            @param timeout as double. Timeout to consider that the topic is not receiving data anymore
            @param required as bool. Flag to include this topic when it checks the overall status of the topics
            @return 0 if ok
            @return -1 if error
        '''
        if type(subscriber) is not rospy.Subscriber:
            rospy.logerr('%s::add_topics_health: the object subscribed is not the correct type -> %s',
                         self._node_name, type(subscriber))
            return -1

        _topic_id = ''
        if topic_id != '':
            _topic_id = topic_id
        else:
            _topic_id = subscriber.resolved_name

        if timeout <= 0:
            rospy.logerr('%s::add_topics_health: timeout (%.lf) has to be >= 0. Setting 1.', self._node_name, timeout)
            timeout = 1.0

        self._data_health_monitors[_topic_id] = TopicHealthMonitor(subscriber, timeout, required)
        rospy.loginfo('%s::add_topics_health: Add topic %s', self._node_name, _topic_id)
        return 0

    def tick_topics_health(self, topic_id):
        '''
            @brief Ticks a topic health as it is receiving data
            @return 0 if OK
            @return -1 if the id doesn't exist
        '''
        if topic_id not in self._data_health_monitors:
            rospy.logerr('%s::tick_topics_health: the topic %s does not exist!', self._node_name, topic_id)
            return -1

        self._data_health_monitors[topic_id].tick()

    def check_topics_health(self, topic_id=''):
        '''
            @brief Checks the topic health of all the subscribed topics or specific ones
            @param topic as string, topic to check. If empty all the topics are checked as a group
            @return true if health is ok, false otherwise
        '''
        if len(self._data_health_monitors) == 0:
            rospy.logwarn('%s::check_topics_health: no topics to check!', self._node_name)
            return True

        if topic_id != '':
            if topic_id not in self._data_health_monitors:
                rospy.logerr('%s::check_topics_health: the topic %s does not exist!', self._node_name, topic_id)
                return False
            else:
                return self._data_health_monitors[topic_id].is_receiving()

        ret = True
        topics_not_received = []
        for i in self._data_health_monitors:
            if self._data_health_monitors[i].is_required() and self._data_health_monitors[i].is_receiving() == False:
                ret = False
                topics_not_received.append(i)

        if ret == False:
            rospy.logwarn_throttle(5, '%s::check_topics_health: topic(s) %s not receiving' %
                                   (self._node_name, str(topics_not_received)))

        return ret

    """
    def topic_cb(self, msg):
    	'''
    		Callback for inelfe_video_manager state
    		@param msg: received message
    		@type msg: std_msgs/Int32
    	'''
    	# DEMO
    	rospy.loginfo('RComponent:topic_cb')


    def service_cb(self, req):
    	'''
    		ROS service server
    		@param req: Required action
    		@type req: std_srv/Empty
    	'''
    	# DEMO
    	rospy.loginfo('RComponent:service_cb')
    """


'''
def main():

	rospy.init_node("rcomponent")


	_name = rospy.get_name().replace('/','')

	arg_defaults = {
	  'topic_state': 'state',
	  '_desired_freq': DEFAULT_FREQ,
	}

	args = {}

	for name in arg_defaults:
		try:
			if rospy.search_param(name):
				args[name] = rospy.get_param('~%s'%(name)) # Adding the name of the node, because the para has the namespace of the node
			else:
				args[name] = arg_defaults[name]
			#print name
		except rospy.ROSException as e:
			rospy.logerr('%s: %s'%(e, _name))


	rc_node = RComponent(args)

	rospy.loginfo('%s: starting'%(_name))

	rc_node.start()


if __name__ == "__main__":
	main()
'''
